app/gui/view/graph-editor/src/free_place_finder.rs

```diff
@@ -0,0 +1,257 @@
+        direction:       Vector2,
+        occupied:        Vec<OccupiedArea>,
+        expected_result: Vector2,
+    }
+
+    impl Case {
+        fn run(&self) {
+            let occupied = self.occupied.iter().cloned();
+            let result = find_free_place(self.starting_point, self.direction, occupied).unwrap();
+            assert_eq!(result, self.expected_result, "Case {:?} gave wrong result.", self);
+        }
+
+        fn flip(&self, on_x: bool, on_y: bool) -> Self {
+            let x_factor = if on_x { -1.0 } else { 1.0 };
+            let y_factor = if on_y { -1.0 } else { 1.0 };
+            let factor = Vector2(x_factor, y_factor);
+            Case {
+                starting_point:  self.starting_point.component_mul(&factor),
+                direction:       self.starting_point.component_mul(&factor),
+                occupied:        self
+                    .occupied
+                    .iter()
+                    .map(|area| OccupiedArea {
+                        x1: area.x1 * factor.x,
+                        x2: area.x2 * factor.x,
+                        y1: area.y1 * factor.y,
+                        y2: area.y2 * factor.y,
+                    })
+                    .collect(),
+                expected_result: self.expected_result.component_mul(&factor),
+            }
+        }
+
+        fn run_each_flip(&self) {
+            self.run();
+            self.flip(true, false).run();
+            self.flip(false, true).run();
+            self.flip(true, true).run();
+        }
+    }
+
+    #[test]
+    fn already_in_unoccupied() {
+        // s> - starting point moving right.
+        // X  - occupied areas.
+        //   +---+
+        //   |XXX|
+        // +-+---+-+
+        // |X|   |X|
+        // |X| s>|X|
+        // |X|   |X|
+        // +-+---+-+
+        //   |XXX|
+        //   +---+
+        let case = Case {
+            starting_point:  Vector2(1.0, -1.0),
+            direction:       Vector2(1.0, 0.0),
+            occupied:        vec![
+                OccupiedArea { x1: 2.0, x2: 3.0, y1: 0.0, y2: -2.0 },
+                OccupiedArea { x1: -1.0, x2: 0.0, y1: 0.0, y2: -2.0 },
+                OccupiedArea { x1: 0.0, x2: 2.0, y1: 1.0, y2: 0.0 },
+                OccupiedArea { x1: 0.0, x2: 2.0, y1: -2.0, y2: -3.0 },
+            ],
+            expected_result: Vector2(1.0, -1.0),
+        };
+        case.run_each_flip()
+    }
+
+    #[test]
+    fn single_area() {
+        let orthogonal = Case {
+            starting_point:  Vector2(10.0, 10.0),
+            direction:       Vector2(1.0, 0.0),
+            occupied:        vec![OccupiedArea { x1: 0.0, x2: 20.0, y1: 15.0, y2: 5.0 }],
+            expected_result: Vector2(20.0, 10.0),
+        };
+        let non_orthogonal = Case {
+            direction: Vector2(18.0, 15.0),
+            expected_result: Vector2(18.0, 15.0),
+            ..orthogonal.clone()
+        };
+        orthogonal.run_each_flip();
```

Comment:
Are we sure this test is run by CI and passing? When I selectively run it locally, it appears to fail, as demonstrated in the transcript below. (Note: the patch adding `features = ["serde"]` is a workaround for #3319.)

```console
$ git checkout origin/wip/vitvakatu/fix-regression-181528359
HEAD is now at 278c6c704 Fix regression #181528359

$ patch -p1 <<'EOF'
diff --git a/app/gui/view/graph-editor/Cargo.toml b/app/gui/view/graph-editor/Cargo.toml
index 760d8b24b..07fb743d9 100644
--- a/app/gui/view/graph-editor/Cargo.toml
+++ b/app/gui/view/graph-editor/Cargo.toml
@@ -15,7 +15,7 @@ bimap = { version = "0.4.0" }
 enso-config = { version = "0.1.0", path = "../../config" }
 enso-frp = { version = "0.1.0", path = "../../../../lib/rust/frp" }
 enso-logger = { path = "../../../../lib/rust/logger"}
-enso-prelude = { path = "../../../../lib/rust/prelude"}
+enso-prelude = { path = "../../../../lib/rust/prelude", features = ["serde"] }
 engine-protocol = { version = "0.1.0", path = "../../controller/engine-protocol" }
 enso-shapely = { path = "../../../../lib/rust/shapely"}
 enso-text = { version = "0.1.0", path = "../../../../lib/rust/text" }
EOF
patching file app/gui/view/graph-editor/Cargo.toml

$ cargo test free_place_finder --manifest-path=app/gui/view/graph-editor/Cargo.toml
(...)
running 4 tests
test free_place_finder::tests::zero_direction_is_not_allowed ... ok
test free_place_finder::tests::already_in_unoccupied ... ok
test free_place_finder::tests::overlapping_areas ... ok
test free_place_finder::tests::single_area ... FAILED

failures:

---- free_place_finder::tests::single_area stdout ----
thread 'free_place_finder::tests::single_area' panicked at 'assertion failed: `(left == right)`
  left: `Matrix { data: [[-15.0, 15.0]] }`,
 right: `Matrix { data: [[-20.0, 10.0]] }`: Case Case { starting_point: Matrix { data: [[-10.0, 10.0]] }, direction: Matrix { data: [[-10.0, 10.0]] }, occupied: [OccupiedArea { x1: -0.0, y1: 15.0, x2: -20.0, y2: 5.0 }], expected_result: Matrix { data: [[-20.0, 10.0]] } } gave wrong result.', app/gui/view/graph-editor/src/free_place_finder.rs:163:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    free_place_finder::tests::single_area
```